google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/JsonStreamWriter.java

```diff
@@ -0,0 +1,377 @@
+      builder.setExecutorProvider(executorProvider);
+    }
+    if (endpoint != null) {
+      builder.setEndpoint(endpoint);
+    }
+    if (onSchemaUpdateRunnable != null) {
+      onSchemaUpdateRunnable.setJsonStreamWriter(this);
+      builder.setOnSchemaUpdateRunnable(onSchemaUpdateRunnable);
+    }
+  }
+
+  /**
+   * Setter for descriptor.
+   *
+   * @param tableSchema
+   */
+  public synchronized void setDescriptor(Table.TableSchema tableSchema)
+      throws Descriptors.DescriptorValidationException {
+    this.descriptor =
+        BQTableSchemaToProtoDescriptor.convertBQTableSchemaToProtoDescriptor(tableSchema);
+  }
+
+  /**
+   * newBuilder that constructs a JsonStreamWriter builder with BigQuery client being initialized by
+   * StreamWriter by default.
+   *
+   * @param streamName name of the stream that must follow
+   *     "projects/[^/]+/datasets/[^/]+/tables/[^/]+/streams/[^/]+"
+   * @param tableSchema The schema of the table when the stream was created, which is passed back
+   *     through {@code WriteStream}
+   * @return Builder
+   */
+  public static Builder newBuilder(String streamName, Table.TableSchema tableSchema) {
+    Preconditions.checkNotNull(streamName, "StreamName is null.");
+    Preconditions.checkNotNull(tableSchema, "TableSchema is null.");
+    return new Builder(streamName, tableSchema, null);
+  }
+
+  /**
+   * newBuilder that constructs a JsonStreamWriter builder.
+   *
+   * @param streamName name of the stream that must follow
+   *     "projects/[^/]+/datasets/[^/]+/tables/[^/]+/streams/[^/]+"
+   * @param tableSchema The schema of the table when the stream was created, which is passed back
+   *     through {@code WriteStream}
+   * @param client
+   * @return Builder
+   */
+  public static Builder newBuilder(
+      String streamName, Table.TableSchema tableSchema, BigQueryWriteClient client) {
+    Preconditions.checkNotNull(streamName, "StreamName is null.");
+    Preconditions.checkNotNull(tableSchema, "TableSchema is null.");
+    Preconditions.checkNotNull(client, "BigQuery client is null.");
+    return new Builder(streamName, tableSchema, client);
+  }
+
+  /** Closes the underlying StreamWriter. */
+  @Override
+  public void close() {
+    this.streamWriter.close();
+  }
+
+  public static final class Builder {
+    private String streamName;
+    private BigQueryWriteClient client;
+    private Table.TableSchema tableSchema;
+
+    private TransportChannelProvider channelProvider;
+    private CredentialsProvider credentialsProvider;
+    private BatchingSettings batchingSettings;
+    private RetrySettings retrySettings;
+    private ExecutorProvider executorProvider;
+    private String endpoint;
+    private OnSchemaUpdateRunnable onSchemaUpdateRunnable;
+
+    private final OnSchemaUpdateRunnable ON_SCHEMA_UPDATE_RUNNABLE =
+        new OnSchemaUpdateRunnable() {
+          public void run() {
+            try {
+              this.getStreamWriter().refreshAppend();
```

Comment:
You need to make refreshAppend and setDescriptor go together, preventing any append actions in between. Because if there is any Append in between, then the setDescriptor is too late.

Reply:
Good catch. I'll change this.